### PR TITLE
allow configuring delayed closed timeout

### PR DESCRIPTION
**Pull request by ramaraochavali**
The Envoy default delayed close timeout of `1s` is not sufficient for many use cases. This PR makes it configurable and defaults to the current value.

Fixes
https://github.com/istio/istio/issues/30577
https://github.com/istio/istio/issues/29172

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 1]
```diff
@@ -379,27 +379,6 @@ func (configgen *ConfigGeneratorImpl) createGatewayHTTPFilterChainOpts(node *mod
...
-		if proxyConfig.GatewayTopology.ForwardClientCertDetails != meshconfig.Topology_UNDEFINED {
-			forwardClientCertDetails = util.MeshConfigToEnvoyForwardClientCertDetails(proxyConfig.GatewayTopology.ForwardClientCertDetails)
-		}
-	}
-
-	var stripPortMode *hcm.HttpConnectionManager_StripAnyHostPort = nil
-	if features.StripHostPort {
-		stripPortMode = &hcm.HttpConnectionManager_StripAnyHostPort{StripAnyHostPort: true}
-	}
-
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:402]: Would suggest not weaving these unrelated changes

**Review by hzxuzhonghu — COMMENTED** [review]
The delayedCloseTimeout change looks good to me

File: pilot/pkg/networking/core/v1alpha3/gateway.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/gateway.go:402]: This was added because we are building http connection manager twice and I had to set that option at both places. So I have refactored it to single function so that I can set once. If you think it is better to do it in a separate PR - I am fine, but it is some what related to this change.

**Review by howardjohn — CHANGES_REQUESTED** [review]
I am not sure add an environment variable for this is correct. This is trivial to set with an Envoyfilter, so I would expect we either make a first class API or use EnvoyFilter?

**Comment by ramaraochavali** [comment]
> I am not sure add an environment variable for this is correct. This is trivial to set with an Envoyfilter, so I would expect we either make a first class API or use EnvoyFilter?

I agree.  I prefer an API in MeshConfig similar to connect_timeout as this is going to be needed for many people. If you are fine with that  - I can make API change.

**Comment by ramaraochavali** [comment]
@howardjohn WDYT? Should I make API change in mesh config?

**Comment by ramaraochavali** [comment]
@howardjohn WDYT ^^?

**Comment by howardjohn** [comment]
I think mesh config is the right place if we are going to add it to an API.
But I would need to investigate a bit more to see how common this is, when
users should configure it, etc

**Comment by ramaraochavali** [comment]
Sure. Please take your time and let me know. This has been a constant request/question in Envoy and found few issues in Istio as well (See the issues listed above). The Envoy default is not suitable for many like websockets etc.

**Comment by howardjohn** [comment]
so the idea is to set this to 0s? Or to something greater than 1s?

**Comment by ramaraochavali** [comment]
The idea is to default to current `1s` so that people do not find surprising behaviours but document this so that people can change it as needed either disable (0s - which might have some implications with rogue clients) or set higher than 1s like 5s depending on their use case. However I do think we can even set to 5s as default to start with.

**Comment by ramaraochavali** [comment]
@howardjohn Did you get a chance to think about this?

**Comment by lambdai** [comment]
LGTM.
Some dumb http 1 client might be impacted by this delay close behavior: those relies on the socket close to determine the end of the request.
But an env var doesn't hurt.

I prefer we use env var in the current version before we promote it to mesh config.a

**Comment by ramaraochavali** [comment]
@lambdai Thank you. I am fine either way leaving it in env var (that was my original intent as well to test this and then promote to API) or mesh config.

@howardjohn @hzxuzhonghu WDYT?

**Review by howardjohn — APPROVED** [review]
